internal/servers/worker/internal/metric/proxy_ws_test.go

```diff
@@ -0,0 +1,450 @@
+			wantPusher:  true,
+		},
+		{
+			name:        "flusher and pusher",
+			wrap:        flusherPusher{},
+			wantFlusher: true,
+			wantPusher:  true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			wrapped := wrapProxyHandler(&hijackerWrapper{}, tt.wrap)
+			require.NotNil(t, wrapped)
+
+			// The wrapped interface should always be a Hijacker and ResponseWriter.
+			require.Implements(t, (*http.ResponseWriter)(nil), wrapped)
+			require.Implements(t, (*http.Hijacker)(nil), wrapped)
+
+			if tt.wantFlusher {
+				require.Implements(t, (*http.Flusher)(nil), wrapped)
+			}
+			if tt.wantPusher {
+				require.Implements(t, (*http.Pusher)(nil), wrapped)
+			}
+		})
+	}
+}
+
+func TestHandlerWrapper_WithHijacker(t *testing.T) {
+	handlerCalled := false
+	h := http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) { // Mock for the proxy handler.
+		handlerCalled = true
+
+		require.NotNil(t, req)
+		require.NotNil(t, rw)
+
+		// We only wrap the response writer in metrics code if it implements Hijacker.
+		require.IsType(t, &hijackerWrapper{}, rw)
+
+		mw := rw.(*hijackerWrapper)
+		require.NotNil(t, mw.Hijacker)
+		require.NotNil(t, mw.ResponseWriter)
+	})
+
+	rw := &testHttpHijacker{ResponseWriter: httptest.NewRecorder(), t: t}
+	req := httptest.NewRequest("GET", "/v1/proxy", nil)
+	wrapped := InstrumentWebsocketWrapper(h)
+	wrapped.ServeHTTP(rw, req)
+
+	require.True(t, handlerCalled)
+}
+
+func TestHandlerWrapper_WithNoHijacker(t *testing.T) {
+	handlerCalled := false
+	h := http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) { // Mock for the proxy handler.
+		handlerCalled = true
+
+		require.NotNil(t, req)
+		require.NotNil(t, rw)
+		require.IsType(t, &httptest.ResponseRecorder{}, rw) // The websocket metrics code must not have wrapped `rw`
+	})
+
+	rw := httptest.NewRecorder()
+	req := httptest.NewRequest("GET", "/v1/proxy", nil)
+	wrapped := InstrumentWebsocketWrapper(h)
+	wrapped.ServeHTTP(rw, req)
+
+	require.True(t, handlerCalled)
+}
+
+func TestHijack(t *testing.T) {
+	ogConnCount := activeConnCount
+	t.Cleanup(func() { activeConnCount = ogConnCount })
+
+	tpg := &testPrometheusGauge{t: t}
+	activeConnCount = tpg
+
+	underlyingConn := &testConn{t: t, readEOF: true, retWriteLen: true}
+	mw := &hijackerWrapper{Hijacker: &testHttpHijacker{t: t, retConn: underlyingConn}}
+
+	c, brw, err := mw.Hijack()
+	require.NoError(t, err)
+	require.NotNil(t, c)
+	require.NotNil(t, brw)
+
+	require.Equal(t, 1, tpg.incCalledN)
+	require.Equal(t, 0, tpg.decCalledN)
+	require.IsType(t, &wsConn{}, c)
+
+	wsConn := c.(*wsConn)
+	require.Equal(t, underlyingConn, wsConn.Conn) // Underlying conn must be our testConn.
+	require.Equal(t, &mw.dec, wsConn.Once)        // wsConn.Once must be set to a ptr to `dec`.
+
+	// Assert that the buffer Reader and Writer are backed up
+	// by the underlying Conn.
+	_, err = io.ReadAll(brw.Reader)
+	require.NoError(t, err)
+	require.True(t, underlyingConn.readCalled)
```

Comment:
In addition to read being called on the underlying connection, we want to make sure read is called on wsConn or else we won't be measuring what we should be.   Same thing for write.

Reply:
I'm unsure how we'd test that `wsConn.Read` has been called without having test fields in `wsConn`, since it's the object under test and thus cannot be mocked.

Reply:
Addressed [5954753](https://github.com/hashicorp/boundary/pull/1963/commits/59547537e4a5be499d98f5d521b102ab2e057702)